### PR TITLE
Improve performance of transformations

### PR DESCRIPTION
**Pull request by aseyboldt**
I used this code to compare the performance:

```python
import pymc3 as pm
import numpy as np

N = 10000

with pm.Model() as model1:
    pm.Uniform('unif', lower=0, upper=1, shape=N)
    step1 = pm.step_methods.hmc.base_hmc.BaseHMC(
        scaling=np.ones(N), profile=True, use_single_leapfrog=True)

with pm.Model() as model2:
    pm.HalfNormal('unif', sd=1, shape=N)
    step2 = pm.step_methods.hmc.base_hmc.BaseHMC(
        scaling=np.ones(N), profile=True, use_single_leapfrog=True)

np.random.seed(42)
q = np.random.randn(N)
p = np.random.randn(N)
epsilon = np.array(0.01)
for model, step in zip([model1, model2], [step1, step2]):
    q_grad = model.dlogp_array(q)
    %timeit _ = step.leapfrog(q, p, q_grad, epsilon)
```

before:
100 loops, best of 3: 2.27 ms per loop
1000 loops, best of 3: 989 µs per loop

after:
1000 loops, best of 3: 1.34 ms per loop
1000 loops, best of 3: 574 µs per loop

There is still some room for improvement: The implementation of the interval transform computes `sigmoid(y)` and `softmax(-y)`. Both usually compute `exp(-y)`, but theano can't figure out that it could reuse the values. If we had an Op, that computes both at the same time we might be able to save another ~200ms:

```python
>>> %timeit np.exp(q)
1000 loops, best of 3: 226 µs per loop
```

### CODE REVIEW

File: pymc3/distributions/transforms.py [hunk 1]
```diff
@@ -116,13 +119,16 @@ def __init__(self, a, b, eps=1e-6):
...
         a, b = self.a, self.b
-        r = (b - a) / (1 + tt.exp(-x)) + a
+        r = (b - a) * tt.nnet.sigmoid(x) + a
         return r
 
     def forward(self, x):
         a, b, e = self.a, self.b, self.eps
-        r = tt.log(tt.maximum((x - a) / tt.maximum(b - x, e), e))
-        return r
+        return tt.log(x - a) - tt.log(b - x)
```
Comment (ColCarroll) [inline pymc3/distributions/transforms.py]: small thing, but there's a class variable `eps` that can be deleted (both here and in `__init__`.

File: pymc3/distributions/transforms.py [hunk 2]
```diff
@@ -116,13 +119,16 @@ def __init__(self, a, b, eps=1e-6):
...
+
+    def jacobian_det(self, x):
```
Comment (ColCarroll) [inline pymc3/distributions/transforms.py]: maybe a note or a reference to show this is correct?  I think we don't do a good job of documenting all these computations, but it'd help if we did

**Review by ColCarroll — APPROVED** [review]
looks good to me -- it seems like there was some sort of instability being guarded against, but it also seems like good practice for theano to deal with that.

File: pymc3/distributions/transforms.py (hunk 2 above)
Comment (aseyboldt) [reply pymc3/distributions/transforms.py]: Your right, the problem is, where should we put something like that? Putting some long latex stuff in the docstring is not exactly readable:

# Interval transform


Forward:
$$
\phi: [a, b] \to \textbf{R}\\
x \mapsto \text{logit}(\tfrac{x - a}{b - a}) = \log(x - a) - \log(b - x)
$$

Backward:
$$
\phi^{-1}: y\mapsto a + (b − a) \cdot \text{sigmoid}(y)
$$

Jacobian:
$$
\left| \frac{d}{dy} \phi^{-1} \right|
= \frac{1}{1 + e^{-x}} \cdot \left(1 - \frac{1}{1 + e^{-1}}\right)
= \frac{e^{-x}}{(1 + e^{-x})^2}
$$

**Comment by aseyboldt** [comment]
@ColCarroll When the benchmark page is finished, do you think we could put some benchmarks of leapfrog functions in there? That would help to keep changes to nuts separate from other performance improvements.

File: pymc3/distributions/transforms.py (hunk 1 above)
Comment (aseyboldt) [reply pymc3/distributions/transforms.py]: done

**Comment by ColCarroll** [comment]
Would love to -- I've bookmarked this to add to the suite once it is running.  I'm on vacation for another week, but hopefully _next_ weekend I'll get some hosted benchmarks.

**Comment by ColCarroll** [comment]
As far as documentation, I think the proper place is in the class docs, though @fonnesbeck or @twiecki might have an opinion.  I was going to suggest not worrying about it for this PR, but looks like it is already TeX'd up...

Looks good to merge once tests pass, though

**Comment by aseyboldt** [comment]
Ah, right. Tex is rendered in the docs, I forgot about that. I'll put it in then.

**Comment by aseyboldt** [comment]
I get an error when I try to build the docs:

```
>>> cd docs/source
>>> make html
Notebook error:
TemplateNotFound in notebooks/BEST.ipynb:
nbsphinx-rst
make: *** [Makefile:55: html] Error 
```
The `conf.py` mentions a dir `templates_`, but there is no such dir.

**Comment by twiecki** [comment]
Very cool speed-up @aseyboldt.

**Comment by twiecki** [comment]
I rebuilt the docs, not sure why you had problems. Maybe missing dependencies.
